nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py

```diff
@@ -0,0 +1,255 @@
+                    adapter_module = module.adapter_layer[adapter_key]
+                    state_adapter_key = ':'.join([name, adapter_key])
+                    state_dict_[state_adapter_key] = adapter_module.state_dict()
+
+                module.set_enabled_adapters(enabled=True)
+        return state_dict_
+
+    def load_state_dict(self, state_dict, strict: bool = True):
+        """
+        Loads a state_dict expecting the state_dict to contain key,values 
+        only for the adapter parameters.
+        """
+        for name, module in self.frozen_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    adapter_module = module.adapter_layer[adapter_key]
+                    state_adapter_key = ':'.join([name, adapter_key])
+                    adapter_module.load_state_dict(state_dict[state_adapter_key], strict)
+                module.set_enabled_adapters(enabled=True)
+
+    def setup_optimizer_param_groups(self):
+        """
+        ModelPT override. Optimizer will get self._optimizer_param_groups. 
+        Makes two optimizer param groups, one for the frozen model params
+        and one for the prompt-table/prompt-encoder params. The learning 
+        rate for the frozen model's params will always be zero effectively
+        freezing the model's params but still allowing for the needed gradients
+        to be passed around in pipeline parallel models. The prompt-encoder 
+        and/or prompt table will use the learning rate set by the user. 
+        """
+        self.frozen_model.freeze()  # Freeze the entire model
+        param_groups = {'params': [p for p in self.frozen_model.parameters()]}
+        for _, module in self.frozen_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                module.set_enabled_adapters(enabled=True)
+                module.unfreeze_enabled_adapters()  # selectively unfreeze the adapter modules.
+        self._optimizer_param_groups = [param_groups]
+        logging.info(f'Optimizer groups set:\n{self.frozen_model.summarize()}')
+
+    def get_forward_output_and_loss_func(self):
+        def fwd_output_and_loss_func(batch, model):
+            batch = [x.cuda(non_blocking=True) for x in batch]
+            input_ids, labels, loss_mask, position_ids, attention_mask, taskname_ids = batch
+            output_tensor = model(input_ids, position_ids, attention_mask, taskname_ids, labels, inference=False)
+
+            def loss_func(output_tensor):
+                loss = self.frozen_model.loss_func(loss_mask, output_tensor)
+                reduced_loss = average_losses_across_data_parallel_group([loss])
+                return loss, {'avg': reduced_loss}
+
+            return output_tensor, loss_func
+
+        return fwd_output_and_loss_func
+
+    def training_step(self, batch, batch_idx):
```

Comment:
Is this also different from the training step in the parent class?

Reply:
Just a minor difference re how it handles pipeline parallel mode. I plan to refactor the class hierarchy in a subsequent PR.